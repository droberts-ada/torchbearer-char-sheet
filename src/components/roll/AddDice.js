import React from 'react';
import PropTypes from 'prop-types';

import PlusMinus from '../shared/PlusMinus';
import Toggle from '../shared/Toggle';
import TraitDropdown from './TraitDropdown';

import './styles/AddDice.css';

class AddDice extends React.Component {
  render() {
    let blHeaderStyle = "";
    if (this.props.disabledOptions.natureInstead) {
      blHeaderStyle = "hidden";
    } else if (this.props.disabledOptions.beginnersLuckHeaders) {
      blHeaderStyle = "struck";
    }
    return (
      <ul id="roll-add-dice">
        {
          // TODO only display beginner's luck info and "roll using nature" if character does not have this skill
        }
        <Toggle
          name="Roll using Nature"
          subtext="If you don't have the skill, you can roll with (full) Nature instead of using Beginner's Luck. If the test is outside your Nature, your Nature will be taxed by the margin of failure. If you are Afraid, this option will be selected automatically."
          disabled={this.props.disabledOptions.natureInstead || this.props.disabledOptions.unselectNatureInstead}
          active={this.props.modifiers.natureInstead}
          onToggle={(value) => this.props.onSetProperty('natureInstead', value)}
          className={
            this.props.disabledOptions.natureInstead ? "hidden" : ""
          }
        />
        <li className={blHeaderStyle}>
          <h2>Before Beginner&apos;s Luck</h2>
        </li>
        <PlusMinus
          name="Help"
          subtext="Get help from your allies, through wises or an appropriate skill"
          value={this.props.modifiers.help}
          min={0}
          onValueChange={(help) => this.props.onSetProperty('help', help)}
        />
        <Toggle
          name="Supplies"
          active={this.props.modifiers.supplies}
          onToggle={(value) => this.props.onSetProperty('supplies', value)}
        />
        <Toggle
          name="Gear"
          active={this.props.modifiers.gear}
          onToggle={(value) => this.props.onSetProperty('gear', value)}
        />
        <li className={blHeaderStyle}>
          <h2>After Beginner&apos;s Luck</h2>
        </li>
        <TraitDropdown
          traitName={this.props.modifiers.traitName}
          traitEffect={this.props.modifiers.traitEffect}
          characterTraits={this.props.character.traits}
          disabledOptions={this.props.disabledOptions}
          onSetProperty={this.props.onSetProperty}
        />
        <PlusMinus
          name="Persona Dice"
          subtext="Spend up to three Persona to gain extra dice on any roll"
          value={this.props.modifiers.personaDice}
          min={0}
          max={3}
          onValueChange={(personaDice) => this.props.onSetProperty('personaDice', personaDice)}
        />
        <Toggle
          name="Tap Nature"
          subtext="Spend a Persona to add your Nature to this test. If the test is outside your Nature, Nature will be taxed by one."
          disabled={this.props.disabledOptions.tapNature}
          active={this.props.modifiers.tapNature}
          onToggle={(value) => this.props.onSetProperty('tapNature', value)}
        />
        {
          // TODO: indicator for fresh (non-interactive)
        }
      </ul>
    );
  }
}

AddDice.propTypes = {
  disabledOptions: PropTypes.object.isRequired,
  onSetProperty: PropTypes.func.isRequired,
  info: PropTypes.object.isRequired,
  modifiers: PropTypes.object.isRequired,
  character: PropTypes.object.isRequired,
};

export default AddDice;
