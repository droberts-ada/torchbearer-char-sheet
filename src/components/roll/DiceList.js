import React from 'react';
import PropTypes from 'prop-types';

import _ from 'underscore';

import diceImages from './images/diceImages';

import './styles/DiceList.css';

const DiceList = ({ dice, name, extra }) => {
  let count = dice.length.toString();
  let successDice = "";
  if (extra || extra === 0) {
    count += ' ';
    count += extra >= 0 ? '+' : '-';
    count += ' ' + Math.abs(extra);
    count += ' = ' + (dice.length + extra);
    if (extra > 0) {
      successDice = _.times(extra, i => (
        <li
          className="die-image-container free-success"
          key={'s_' + i}
        >
          <label>S</label>
        </li>
      ))
    }
  }
  return (
    <div className="dice-list">
      <h2>
        {name + ' '}
        ({count})
      </h2>
      <ul>
        {successDice}
        {
          dice.map((die, i) => {
            let className = "die-image-container" + (die.rerolled ? " rerolled" : "");
            if (i >= dice.length + extra) {
              className += " cancelled";
            }
            return (
              <li
                className={className}
                key={die.id}
              >
                <img
                  src={diceImages[die.face]}
                  alt={`die-${die.face}`}
                />
              </li>
            );
          })
        }
      </ul>
    </div>
  );
};

DiceList.propTypes = {
  dice: PropTypes.array.isRequired,
  name: PropTypes.string.isRequired,
  extra: PropTypes.number,
};

export default DiceList;