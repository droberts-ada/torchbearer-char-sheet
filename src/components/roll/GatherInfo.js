import React from 'react';

import Checkbox from '../shared/Checkbox';
import PlusMinus from '../shared/PlusMinus';
import SkillAbilityDropdown from './SkillAbilityDropdown';

const GatherInfo = function(props) {
  return (
    <ul id="roll-gather-info">
      <li>
        <h3>Versus test</h3>
        <Checkbox
          active={props.info.type === 'versus'}
          onToggle={() => {props.onSetVersus(props.info.type !== 'versus')}}
          />
      </li>
      <li>
        <h3>Obstacle</h3>
        <PlusMinus
          value={props.info.ob}
          min={0}
          onValueChange={props.onSetOb}
          />
      </li>
      <li>
        <h3>Skill / Ability</h3>
        <SkillAbilityDropdown
          character={props.character}
          current={props.info.skill}
          onSelectSkill={props.onSelectSkill}
          />
      </li>
    </ul>
  );
}

export default GatherInfo;
