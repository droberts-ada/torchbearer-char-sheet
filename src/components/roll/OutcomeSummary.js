import React from 'react';
import PropTypes from 'prop-types';

import './styles/OutcomeSummary.css';

const OutcomeSummary = ({ outcome, ob, totalSuccesses }) => {
  return (
    <div className="outcome-summary">
      <h2>Outcome: {outcome}</h2>

      <div className="outcome-stats">
        <label htmlFor="ob">
          Obstacle
        </label>
        <label htmlFor="successes">
          Successes
        </label>
        <label htmlFor="margin">
          Margin
        </label>
        <span className="number" name="ob">
          {ob}
        </span>
        <span className="number" name="successes">
          {totalSuccesses}
        </span>
        <span className="number" name="margin">
          {totalSuccesses - ob}
        </span>
      </div>
    </div>
  );
};

OutcomeSummary.propTypes = {
  outcome: PropTypes.string.isRequired,
  ob: PropTypes.number.isRequired,
  totalSuccesses: PropTypes.number.isRequired,
}

export default OutcomeSummary;