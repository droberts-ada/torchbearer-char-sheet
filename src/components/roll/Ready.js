import React from 'react';

import "./styles/Ready.css";

const buildDetail = (detail, i) => {
  return (
    <li
      key={`detail_${i}`}>
      <p>
        <strong>{detail.effect}</strong>
        &nbsp;from&nbsp;
        <em>{detail.source}</em>
        <br></br>
        <span className="reason">{detail.reason}</span>
      </p>
    </li>
  );
}

const Ready = ({ details, rollDice }) => {
  return (
    <div id="roll-ready">
      <ul>
        {details.map(buildDetail)}
      </ul>
      <button
        className="action-button"
        onClick={rollDice}
      >
        Roll the dice!
        </button>
    </div>
  );
}

export default Ready;
