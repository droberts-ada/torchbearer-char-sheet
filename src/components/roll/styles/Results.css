.roll-results .outcome {
  padding: .5em 0;
  border-top: 1px solid black;
  width: 100%;

  position: fixed;
  bottom: 0px;

  background-color: white;
}

.roll-results .outcome h2 {
  font-size: 1.75em;
  text-transform: uppercase;
}

.roll-results .outcome-stats {
  display: grid;
  grid-template-columns: repeat(3, 1fr);
  justify-items: center;
}

.roll-results .outcome button {
  font-size: 1.5em;
  display: block;
  width: 70%;
  padding: .2em;
  margin-left: auto;
  margin-right: auto;
  margin-top: .5em;
}

.roll-results .pre-footer {
  padding-bottom: 3em;
}

.roll-results .reaction-log {
  border-top: 1px solid black;
}

.roll-results .reaction-log .log-line {
  margin: .5em 0;
}