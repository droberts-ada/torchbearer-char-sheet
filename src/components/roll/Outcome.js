import React from 'react';
import PropTypes from 'prop-types';

import pluralize from 'pluralize';

import "./styles/Outcome.css";
import Toggle from '../shared/Toggle';
import OutcomeSummary from './OutcomeSummary';

const Points = ({ name, points, verb = "Spend" }) => {
  return (
    <li key={`points_${name}`}>
      <b>{verb} {points.total[name]}</b> {name}
      <ul className="effect-details">
        {
          points[name].map((effect) => (
            <li key={effect.source}>
              {effect.effect} for <em>{effect.source}</em>
            </li>
          ))
        }
      </ul>
    </li>
  );
}

const BeneficialTrait = ({ trait }) => {
  if (trait) {
    return (
      <li key="beneficialTrait">
        <p>
          <b>Mark</b> beneficial use of trait <em>{trait}</em>
        </p>
      </li>
    );
  }
  return "";
}

const MarkEffect = ({ effect }) => {
  return (
    <li key={`me_${effect.category}_${effect.name}`}>
      <b>Mark</b> use of {pluralize.singular(effect.category)} <em>{effect.name}</em>
      <ul className="effect-details">
        <li>
          Mark type: <em>{effect.mark}</em>
          {effect.markComment ? ` (${effect.markComment})` : ""}
        </li>
        {effect.alreadyMarked ? (<li>(already marked)</li>) : ""}
        {effect.advance ? (<li>Ready to advance!</li>) : ""}
      </ul>
    </li>
  )
}

const Outcome = ({ impact, preRoll, postRoll, operations }) => {
  // Page 65
  let outcomeFlavor;
  if (postRoll.outcome === 'pass') {
    outcomeFlavor = "When you pass an ability or skill test, you get what you want. You should offer a little bit of description to celebrate the moment.";
  } else {
    outcomeFlavor = "If a player fails an ability or skill test, one of two things can happen: either a twist is introduced or the character succeeds, but with a condition."
  }
  return (
    <div className="roll-outcome">
      <OutcomeSummary
        outcome={postRoll.outcome}
        ob={preRoll.summary.ob}
        totalSuccesses={postRoll.totalSuccesses}
      />
      <p>{outcomeFlavor}</p>

      <div>
        <h2>Roll Effects</h2>
        <ul className="impact-effect-list">
          <MarkEffect effect={impact.skill} />
          {impact.wises.map(effect => <MarkEffect effect={effect} />)}
          <BeneficialTrait trait={impact.beneficialTrait} />
          <Points
            name="persona"
            points={impact.points}
          />
          <Points
            name="fate"
            points={impact.points}
          />
          <Points
            name="checks"
            points={impact.points}
            verb="Earn"
          />
        </ul>
      </div>

      <div>
        <h2>Next Steps</h2>
        <ul>
          <Toggle
            name="Skill already used this conflict*"
            className="skill-used-already-checkbox"

          />
          <li>
            <button
              className="action-button"
              onClick={operations.commitResults}
            >
              Apply effects to character
          </button>
          </li>
          <li>
            <button
              className="action-button"
              onClick={operations.reset}
            >
              Reset roll without applying effects
          </button>
          </li>
          <li>
            <p>
              *Any time you test an ability multiple times to determine the outcome, only one test is earned toward advancement. Log the first test you earn. Thatâ€™s the one that counts for this conflict.
          </p><p>
              Exception: If you only need one more of a particular type of test to advance, you can hold off noting your test during a conflict to see if you get the pass or fail that you need to advance.
          </p>
          </li>
        </ul>
      </div>
    </div >
  );
};

Outcome.propTypes = {
  impact: PropTypes.object.isRequired,
  preRoll: PropTypes.object.isRequired,
  postRoll: PropTypes.object.isRequired,
  operations: PropTypes.object.isRequired,
}

export default Outcome;