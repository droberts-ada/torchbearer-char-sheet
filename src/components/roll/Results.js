import React from 'react';
import PropTypes from 'prop-types';

import './styles/Results.css';
import Control from '../shared/Control';
import DiceList from './DiceList';
import OutcomeSummary from './OutcomeSummary';

const UseButton = ({ name, onSetReaction, disabled = false }) => {
  return (
    <button
      disabled={disabled}
      onClick={() => onSetReaction(name, true)}
      className="use-button"
    >
      Use
    </button>
  )
};

const WiseList = ({ characterWises, selectedWise = "none", onSelectWise, disabled, disabledWises }) => {
  return (
    <select
      value={selectedWise}
      onChange={event => onSelectWise(event.target.value)}
      disabled={disabled}
    >
      <option key="none" value="none">
        Wise - None
      </option>
      {
        characterWises.map(wise => {
          return (
            <option
              key={`wise_${wise.name}`}
              value={wise.name}
              disabled={disabledWises.includes(wise.name)}
            >
              {wise.name}
            </option>
          );
        })
      }
    </select>
  );
};

WiseList.propTypes = {
  characterWises: PropTypes.array.isRequired,
  selectedWise: PropTypes.string,
  onSelectWise: PropTypes.func.isRequired,
  disabled: PropTypes.bool,
  disabledWises: PropTypes.array.isRequired,
}

const WiseReaction = ({ name, subtext, reactionName, reactions, characterWises, onSetReaction, disabledOptions }) => {
  const selectedWise = reactions[reactionName + 'Wise'];
  return (
    <Control
      className="roll-reaction wise-reaction"
      name={name}
      subtext={subtext}
      knob={(
        <React.Fragment>
          <UseButton
            name={reactionName + 'Used'}
            onSetReaction={onSetReaction}
            disabled={disabledOptions.button}
          />
          <WiseList
            characterWises={characterWises}
            selectedWise={selectedWise}
            onSelectWise={wise => {
              onSetReaction(reactionName + 'Wise', wise)
            }}
            disabled={disabledOptions.select}
            disabledWises={disabledOptions.wises}
          />

        </React.Fragment>
      )}
    />
  );
};

// TODO: consider changing name to OutcomePreview
const OutcomePreview = ({ postRoll, onAcceptRoll }) => {
  return (
    <footer className="outcome">
      <OutcomeSummary
        postRoll={postRoll}
        />

      <button
        className="action-button"
        onClick={onAcceptRoll}
      >
        Accept Results
      </button>
    </footer>
  )
};

const Results = ({ postRoll, reactions, rollSummary, character, onSetReaction, onAcceptRoll, disabledOptions, resourcesSpent }) => {
  const { successes, scoundrels } = postRoll;

  const availableFate = character.points.fate.available - resourcesSpent.fate;
  const availablePersona = character.points.persona.available - resourcesSpent.persona;

  return (
    <div
      className="roll-results"
    >
      <section className="pre-footer">
        <DiceList
          dice={successes}
          name="Successes"
          extra={rollSummary.addSuccesses} />
        <DiceList
          dice={scoundrels}
          name="Scoundrels" />
        <h2>Reactions</h2>
        <ul>
          <WiseReaction
            name="Deeper Understanding"
            subtext={(
              <React.Fragment>
                Spend a fate point and reroll any single failed die on a test related to your wise. State, “Ah hah!” and gesture that you understand everything now.
                <br /><br />
                <strong>
                  {availableFate}
                  &nbsp;fate available
                </strong>
              </React.Fragment>
            )}
            reactionName="deeperUnderstanding"
            reactions={reactions}
            characterWises={character.wises}
            onSetReaction={onSetReaction}
            disabledOptions={disabledOptions.deeperUnderstanding}
          />
          <WiseReaction
            name="Of Course"
            subtext={(
              <React.Fragment>
                Spend a persona point and reroll all failed dice on a test related to your wise. Declare, “Of course!” and indicate that you were wrong before but you have it all correct now.
              <br /><br />
                <strong>
                  {availablePersona}
                  &nbsp;persona available
                </strong>
              </React.Fragment>
            )}
            reactionName="ofCourse"
            reactions={reactions}
            characterWises={character.wises}
            onSetReaction={onSetReaction}
            disabledOptions={disabledOptions.ofCourse}
          />
          <Control
            className="roll-reaction"
            name="Fate for Luck"
            subtext={(
              <React.Fragment>
                Spend a fate point and reroll all sixes. Sixes rolled this way keep rerolling. Describe the Luck!
              <br /><br />
                <strong>
                  {availableFate}
                  &nbsp;fate available
                </strong>
              </React.Fragment>
            )}
            knob={(
              <UseButton
                name="explodeSixes"
                disabled={disabledOptions.explodeSixes}
                onSetReaction={onSetReaction}
              />
            )}
          />
          {/* TODO: tiebreakers */}
          <section className="reaction-log">
            <h3>Log</h3>
            <ul>{
              reactions.log.map((logLine, i) => (
                <li className="log-line" key={i}><p>{logLine}</p></li>
              ))
            }</ul>
          </section>
        </ul>
      </section>
      <OutcomePreview
        postRoll={postRoll}
        onAcceptRoll={onAcceptRoll}
      />
    </div>
  );
};

Results.propTypes = {
  rolledDice: PropTypes.array.isRequired,
  reactions: PropTypes.object.isRequired,
  rollSummary: PropTypes.object.isRequired,
  character: PropTypes.object.isRequired,
  onSetReaction: PropTypes.func.isRequired,
  onAcceptRoll: PropTypes.func.isRequired,
  disabledOptions: PropTypes.object.isRequired,
  resourcesSpent: PropTypes.object.isRequired,
};

export default Results;