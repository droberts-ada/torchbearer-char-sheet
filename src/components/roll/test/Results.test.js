import React from 'react';
import renderer from 'react-test-renderer';

import Results from '../Results';

import character from '../../../mock/character';
import roll from '../../../mock/roll';
import calculateDerivedRollState from '../../../derivers/roll/derived_state';

describe('Roll Results page', () => {
  it('renders with a typical roll / character setup', () => {
    const tree = renderer
      .create(
        <Results
          rolledDice={roll.results.rolledDice}
          reactions={roll.results.reactions}
          rollSummary={calculateDerivedRollState(roll, character)}
          characterWises={character.wises}
          onSetReaction={() => {}}
          />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
})

