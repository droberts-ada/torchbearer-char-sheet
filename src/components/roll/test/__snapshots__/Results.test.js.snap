// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Roll Results page renders with a typical roll / character setup 1`] = `
<div
  className="roll-results"
>
  <section
    className="pre-footer"
  >
    <div
      className="dice-list"
    >
      <h2>
        Successes 
        (
        5
        )
      </h2>
      <ul>
        
        <li
          className="die-image-container"
        >
          <img
            alt="die-6"
            src="dice-6.png"
          />
        </li>
        <li
          className="die-image-container"
        >
          <img
            alt="die-6"
            src="dice-6.png"
          />
        </li>
        <li
          className="die-image-container"
        >
          <img
            alt="die-5"
            src="dice-5.png"
          />
        </li>
        <li
          className="die-image-container"
        >
          <img
            alt="die-5"
            src="dice-5.png"
          />
        </li>
        <li
          className="die-image-container"
        >
          <img
            alt="die-4"
            src="dice-4.png"
          />
        </li>
      </ul>
    </div>
    <div
      className="dice-list"
    >
      <h2>
        Scoundrels 
        (
        4
        )
      </h2>
      <ul>
        
        <li
          className="die-image-container"
        >
          <img
            alt="die-2"
            src="dice-2.png"
          />
        </li>
        <li
          className="die-image-container"
        >
          <img
            alt="die-2"
            src="dice-2.png"
          />
        </li>
        <li
          className="die-image-container"
        >
          <img
            alt="die-2"
            src="dice-2.png"
          />
        </li>
        <li
          className="die-image-container"
        >
          <img
            alt="die-1"
            src="dice-1.png"
          />
        </li>
      </ul>
    </div>
    <h2>
      Reactions
    </h2>
    <ul>
      <li
        className="roll-reaction wise-reaction control"
      >
        <div
          className="control-main"
        >
          <h3
            className="control-title"
          >
            Deeper Understanding
          </h3>
          <button
            className="use-button"
            disabled={true}
            onClick={[Function]}
          >
            Use
          </button>
          <select
            disabled={false}
            onChange={[Function]}
            value="none"
          >
            <option
              value="none"
            >
              Wise - None
            </option>
            <option
              disabled={false}
              value="Deepest Heart of Darkness"
            >
              Deepest Heart of Darkness
            </option>
            <option
              disabled={false}
              value="Home"
            >
              Home
            </option>
          </select>
        </div>
        <p>
          Spend a fate point and reroll any single failed die on a test related to your wise. State, “Ah hah!” and gesture that you understand everything now.
        </p>
      </li>
      <li
        className="roll-reaction wise-reaction control"
      >
        <div
          className="control-main"
        >
          <h3
            className="control-title"
          >
            Of Course
          </h3>
          <button
            className="use-button"
            disabled={true}
            onClick={[Function]}
          >
            Use
          </button>
          <select
            disabled={false}
            onChange={[Function]}
            value="none"
          >
            <option
              value="none"
            >
              Wise - None
            </option>
            <option
              disabled={false}
              value="Deepest Heart of Darkness"
            >
              Deepest Heart of Darkness
            </option>
            <option
              disabled={false}
              value="Home"
            >
              Home
            </option>
          </select>
        </div>
        <p>
          Spend a persona point and reroll all failed dice on a test related to your wise. Declare, “Of course!” and indicate that you were wrong before but you have it all correct now.
        </p>
      </li>
      <li
        className="roll-reaction control"
      >
        <div
          className="control-main"
        >
          <h3
            className="control-title"
          >
            Fate for Luck
          </h3>
          <button
            className="use-button"
            disabled={false}
            onClick={[Function]}
          >
            Use
          </button>
        </div>
        <p>
          Spend one fate to reroll all sixes. Sixes rolled this way keep rerolling. Describe the Luck!
        </p>
      </li>
      <section
        className="reaction-log"
      >
        <h3>
          Log
        </h3>
        <ul />
      </section>
    </ul>
  </section>
  <footer
    className="outcome"
  >
    <h2>
      Outcome: 
      failure
    </h2>
    <div
      className="outcome-stats"
    >
      <label
        htmlFor="ob"
      >
        Obstacle
      </label>
      <label
        htmlFor="successes"
      >
        Successes
      </label>
      <label
        htmlFor="margin"
      >
        Margin
      </label>
      <span
        className="number"
        name="ob"
      />
      <span
        className="number"
        name="successes"
      >
        NaN
      </span>
      <span
        className="number"
        name="margin"
      >
        NaN
      </span>
    </div>
    <button>
      Accept Results
    </button>
  </footer>
</div>
`;
